### PR TITLE
# Docs-Chapters rearranged

### PR DESCRIPTION
**Pull request by kolibril13**
### 1
We still have to decide where the corresponding examples for the python code in the markdown files will go. @leotrs: You had the idea of adding them into the build/_static folder. In this case, we could make a single folder for each markdown file in examples.

 Or we could have the markdown and the video files closer to each other, e.g. to have a folder for each example, where you have the markdown and the video file. This close-together option would be my preferred option when we make the examples manually. 
 
 When we would make the examples automated (e.g. with the script I wrote in #285) I would prefer the separated folder option.
 
 ### 2
The new folder structure looks like this:
stru

Everything starts at `index.rst` and then 
* the chapters "installation, examples and tutorials" have their own sub-directory, with their own index files, e.g. `installation_index.rst`.  These sub-index files link to other documents, which will be in markdown. Here a question: Is it possible/wanted to replace these sub-index files also with markdown files? (e.g. `installation_index.rst` to `installation_index.md`)
* the chapter contributing does not have its own subdirectory yet and I don't think this is necessary, do you agree?
* The chapter reference also does not have its own subdirectory, I don't know how to deal with these auto-generated folders, maybe this is something you can have a look on, @leotrs? At the moment it throws 281 warnings every time I run it.

 
 ### 3 
 
 In `index.rst` I changed the toctree parameter `:maxdepth:`  from 1 to 2
 
 ### 4 
 
 Lots of content that you currently see is only placeholder content and can be changed anytime.
 
 ### 5
 
 From time to time, I had to delete everything in the build folder and rebuild everything, because there were old pages that strangely interacted with the current pages.

### CODE REVIEW

**Comment by leotrs** [comment]
> ### 1
> We still have to decide where the corresponding examples for the python code in the markdown files will go.

I'm neutral as to these organization, as long as (i) the organization makes sense, and (ii), the files are easy to generate. Either option makes sense to me, and your script should make it easy to generate the videos/files. Finally, my usual stance is to "try something quick now, and modify later if necessary". What's the easiest option that satisfies conditions (i) and (ii)? Let's to that one for now. If it stops making sense in the future, we change it in the future.

I will say however that `buid/_static` is the default/preferred place to put static files (such as videos). But we can choose something else if it makes more sense for us. Perhaps a subfolder inside `build/_static`.


> ### 2
> The new folder structure looks like this:
> ...
> 
> Everything starts at `index.rst` and then

Makes sense to me.

> Here a question: Is it possible/wanted to replace these sub-index files also with markdown files? (e.g. `installation_index.rst` to `installation_index.md`)

I'm trying to keep as much as possible in .rst format. Also these indices, won't be hard to write in rst as they won't need a lot of features, they should probably just contain a short blurb and a list of links. Let's use rst.

> * the chapter contributing does not have its own subdirectory yet and I don't think this is necessary, do you agree?

Don't think it's necessary.

> * The chapter reference also does not have its own subdirectory.

I will create one soon (TM). Leave it as it is for now please.

> At the moment it throws 281 warnings every time I run it.

Yeah, sorry. We'll have to overlook those for the time being, until we populate more and more docstrings across the codebase.


> ### 3
> In `index.rst` I changed the toctree parameter `:maxdepth:` from 1 to 2

SGTM. This is also something I think may or may not change frequently in the future as we change the theme and other organization.
 

> ### 5
> From time to time, I had to delete everything in the build folder and rebuild everything, because there were old pages that strangely interacted with the current pages.

Same here. I think that's just the nature of sphinx. For future reference, for anyone who's reading, you have to execute `make clean` (and sometimes manually delete some old files too) before your `make html` is successful.

File: docs/source/changelog.rst [hunk 1]
```diff
@@ -1,4 +0,0 @@
-Changelog
-=========
-
-This is a stub file.
```
Comment (leotrs) [inline docs/source/changelog.rst:4]: I just realized that the changelog got deleted. I'm strongly in favor of keeping it somewhere in the documentation, as a separate chapter. Can we put it back please? (I know this is just a stub file, but still).

Comment (kolibril13) [reply docs/source/changelog.rst:4]: Yes, it will come back! Lets discuss this in #290

**Comment by kolibril13** [comment]
> I will say however that `buid/_static` is the default/preferred place to put static files (such as videos). But we can choose something else if it makes more sense for us. Perhaps a subfolder inside `build/_static`.
> 
> >
Ok, lets do it this way. I will add a _static folder, and folders for each markdown file for example.



> I'm trying to keep as much as possible in .rst format. Also these indices, won't be hard to write in rst as they won't need a lot of features, they should probably just contain a short blurb and a list of links. Let's use rst.
> 
> >

ok!

**Comment by kolibril13** [comment]
Further point: do we really want to add the gifs to the repo?
Also, when gifs are changed, then the previous gifs will still sit in the git history
When we have lots of examples, this can mean that the repo could become of a size of over 300 mb...

**Comment by Aathish04** [comment]
> Further point: do we really want to add the gifs to the repo?
> Also, when gifs are changed, then the previous gifs will still sit in the git history
> When we have lots of examples, this can mean that the repo could become of a size of over 300 mb...

I know it might seem unprofessional, but could we use external image cdns say like, giphy or imgur?

**Comment by kolibril13** [comment]
> I know it might seem unprofessional, but could we use external image cdns say like, giphy or imgur?

My idea would be to use a separate repo for that:
https://github.com/ManimCommunity/manim-docs-files
Is that ok?
If not I can also delete this again.

**Comment by leotrs** [comment]
> I know it might seem unprofessional, but could we use external image cdns say like, giphy or imgur?

What would this look like?

1. I write a documentation file
2. Generate some gifs
3. Upload to the cloud
4. Link to them from the file
5. Push the file

Like this? Is there a way to have like a shared manim account on these sites? So we can keep all the gifs in one single cloud-folder? Or will they just be assigned random URLs each time?


> My idea would be to use a separate repo for that

We would still need to host each gif somewhere so we can like to them from each documentation file. Or do you suggest we just use GitHub as hosting service?

**Comment by naveen521kk** [comment]
Add to GitHub and use jsdelivr as a cdn. :)

**Comment by kolibril13** [comment]
> Or do you suggest we just use GitHub as hosting service?
yes. 
This workflow could then also easily be automated

**Comment by leotrs** [comment]
Hmmm ok I think I'm onboard with that, but I'd still like to hear from more devs. This is a fairly major change to the face of the organization since it will create a whole new repository. Thoughts from @ManimCommunity/core?

**Comment by naveen521kk** [comment]
I suggest using the same repo and not create a new one. As it would be easy for a user to clone manim once and have docs included which is **batteries included**. So I consider using the same repo instead of creating a new one, if needed we can use this repo for gh-pages(by creating a new branch) or if we do by readthedocs docs repo isn't required. I suggest using jsdelivr instead of just using raw images from GitHub.

**Comment by leotrs** [comment]
> I suggest using the same repo and not create a new one. As it would be easy for a user to clone manim once and have docs included which is batteries included.

I agree, but not everyone who clones the repo will be working on documentation. The gifs included in the docs aren't really necessary for everyone.

> So I consider using the same repo instead of creating a new one,

The other problem is that in the future we will probably have tons of little gifs for documentations and examples and it will bloat the repository by a lot. And git doesn't handle binary files very well. It will store every single past version of each gif...

> I suggest using jsdelivr instead of just using raw images from GitHub.

I'm unfamiliar with jsdelivr. Can you please show an example of how it would integrate with .md or .rst documentation files?

**Comment by kolibril13** [comment]
@naveen521kk: **batteries included** is a good point. 
I think here we must deviate between user convenience and keeping the repo small in size.
The users will have access to the docs online, and the docstrings will be even available locally in this repo.
Only the images/videos will be in another place.
In my early manim days I added my videos to the repo, and even after deleting them my repo had  a very huge size(>1GB), and let's say we want to re-render videos because e.g. self.play() changed, this would multiply the size..
So I would really like to outsource the videos and images to another place.

**Comment by naveen521kk** [comment]
*Maybe* we can create a new repo for gifs and pictures. 
> I'm unfamiliar with jsdelivr. Can you please show an example of how it would integrate with .md or .rst documentation files?

The link below points to jsdelivr cdn for a picture in readme.
https://cdn.jsdelivr.net/gh/ManimCommunity/manim/readme-assets/windows_sox.png
We just use it as normally how we do in md.
```md
![images-alt](https://cdn.jsdelivr.net/gh/ManimCommunity/manim/readme-assets/windows_sox.png)
```
This can be done for images, gif but for video I consider using YouTube and embed it.

**Comment by leotrs** [comment]
@naveen521kk thanks for the example. What's the benefit of using jsdelivr instead of hot linking to GitHub, if an image is hosted on GH already?

**Comment by naveen521kk** [comment]
Hot linking GitHub links causes some problems in site speed as it is not actually a cdn. But jsdelivr doesn't have that problem and is quite faster and will always be up( Many companies support it). So even though we keep our images on GitHub we should use jsdelivr.

**Comment by leotrs** [comment]
Got it, thanks! Makes sense to me.

File: docs/source/reference.rst [hunk 2]
```diff
@@ -1,9 +1,10 @@
-Manim Reference
+Reference
 ===============
 
 This reference manual details functions, modules, and objects included in
 Manim, describing what they are and what they do.  For learning how to use
 Manim, see the :doc:`quickstart` and :doc:`tutorial` documents.
+Want to see whats new? Here is the changelock [Note:add link here to changelog.rst]
```
Comment (leotrs) [inline docs/source/reference.rst:7]: what you are looking for is 

```
:doc:`changelog`
```

File: docs/source/tutorials/quickstart.md [hunk 3]
```diff
@@ -61,8 +62,8 @@ $ manim scene.py SquareToCircle -pl
 After some output, `manim` should render the scene into a .mp4 file, and open
 that file with the default movie player application.  You should see a video
 playing the following animation.
-
-![first scene output](_static/quickstart/first_scene.gif)
+/home/jan-hendrik/projects/manim-community/docs/source/_static/quickstart/first_scene.gif
```
Comment (leotrs) [inline docs/source/tutorials/quickstart.md:65]: Why is this URL here?

**Comment by Aathish04** [comment]
> Hot linking GitHub links causes some problems in site speed as it is not actually a cdn. But jsdelivr doesn't have that problem and is quite faster and will always be up( Many companies support it). So even though we keep our images on GitHub we should use jsdelivr.

I think we can combine @kolibril13's suggestion along with @naveen521kk's suggestion; have the media resources _not_ in this repository, but keep them on a separate repo, and use jsdelivr links in the actual documentation.

Keeping media in the main repository would make it get _really_ bulky _really_ quick, and if we'll have examples for each and every type of animation, for each and every type of object, it could easily hit the hundreds of megabytes range.

**Comment by leotrs** [comment]
> I think we can combine @kolibril13's suggestion along with @naveen521kk's suggestion; have the media resources not in this repository, but keep them on a separate repo, and use jsdelivr links in the actual documentation.

Agreed. Awaiting confirmation from more Devs. I see two votes in favor. @kolibril13 @naveen521kk, do you agree with this course of action?

**Comment by naveen521kk** [comment]
Yeah ok. A new repo should be only for images gif and also the main repo should also contains some docs so that users can view it offline.(at least the text). 
:+1: for new repo name `manim-docs`

File: docs/source/reference.rst (hunk 2 above)
Comment (kolibril13) [reply docs/source/reference.rst:7]: Thanks

File: docs/source/tutorials/quickstart.md (hunk 3 above)
Comment (kolibril13) [reply docs/source/tutorials/quickstart.md:65]: sorry, that was only for testing, forgot to remove it.

**Comment by kolibril13** [comment]
> do you agree with this course of action?

yes, that sounds very good to me!



> for new repo name `manim-docs`

Sounds good! I can eighter delete or rename the repo https://github.com/ManimCommunity/manim-docs-files .
Sorry if  I was a bit premature to create this, I was curious if repo cross-links would work smoothly (They do ;))

**Comment by leotrs** [comment]
I'm partial to `manim-docs-files` or `manim-docs-assetts`. However, let's hold off on that until more devs chime in.

**Comment by leotrs** [comment]
I'm merging this now, and moving the discussion on the new repository to a new separate issue.

**Pull request merged**